satpy/readers/utils.py

```diff
@@ -474,3 +474,179 @@ def remove_earthsun_distance_correction(reflectance, utc_date=None):
+
+    .. code-block:: python
+
+        from satpy.readers.utils import CalibrationCoefficientSelector
+
+        coefs = {
+            "nominal": {
+                "ch1": "nominal_ch1",
+                "ch2": "nominal_ch2",
+                "ch3": "nominal_ch3"
+                "ch4": "nominal_ch4"
+                "ch5": "nominal_ch5"
+            },
+            "meirink": {
+                "ch1": "meirink_ch1",
+            },
+            "gsics": {
+                "ch2": "gsics_ch2",
+                # ch3 coefficients are missing
+            }
+        }
+        wishlist = {
+            "ch1": "meirink",
+            ("ch2", "ch3"): "gsics"
+            "ch4": {"mygain": 123},
+        }
+
+    2. Query:
+
+    .. code-block:: python
+
+        >>> s = CalibrationCoefficientSelector(coefs, wishlist)
+        >>> s.get_coefs("ch1")
+        "meirink_ch1"
+        >>> s.get_coefs("ch2")
+        "gsics_ch2"
+        >>> s.get_coefs("ch3")
+        KeyError: 'No gsics calibration coefficients for ch3'
+        >>> s.get_coefs("ch4")
+        {"mygain": 123}
+        >>> s.get_coefs("ch5")
+        "nominal_ch5
+
+    3. Fallback to nominal for ch3:
+
+    .. code-block:: python
+
+        >>> s = CalibrationCoefficientSelector(coefs, wishlist, fallback="nominal")
+        >>> s.get_coefs("ch3")
+        "nominal_ch3"
+    """
+
+    def __init__(self, coefs, wishlist, default="nominal", fallback=None):
+        """Initialize the coefficient selector.
+
+        Args:
+            coefs (dict): One set of calibration coefficients for each
+                calibration mode. The actual coefficients can be of any type
+                (reader-specific).
+            wishlist (str or dict): Desired calibration coefficients. Use a
+                dictionary to specify channel-specific coefficients. Use a
+                string to specify one mode for all channels.
+            default (str): Default coefficients to be used if nothing was
+                specified in the wishlist. Default: "nominal".
+            fallback (str): Fallback coefficients if the desired coefficients
+                are not available for some channel. By default, an exception is
+                raised if coefficients are missing.
+        """
+        if fallback and fallback not in coefs:
+            raise KeyError("No fallback coefficients")
+        self.coefs = coefs
+        self.wishlist = wishlist
+        self.default = default
+        self.fallback = fallback
+        self.parser = _CalibrationCoefficientParser(coefs, default)
+        self.parsed_wishlist = self.parser.parse(wishlist)
+
+    def get_coefs(self, channel):
+        """Get calibration coefficients for the given channel.
+
+        Args:
+            channel (str): Channel name
+        """
+        try:
+            return self.parsed_wishlist[channel]
+        except KeyError:
+            if self.fallback:
+                return self.coefs[self.fallback][channel]
```

Comment:
Should there be a log message that tells the users that fallback coeffs are used?

Reply:
Good suggestion, done